File: modules/lua/test/core/lua/test_observer.lua
-- Observe a service's state transitions through the ServiceController bridge
local core = sihd.core.Core("core", nil)

local ctrl = core:service_ctrl()
assert(ctrl ~= nil)

local states = {}
ctrl:add_observer(function(c)
    states[#states + 1] = c:state()
end)

assert(core:init())
assert(core:start())
assert(core:stop())

ctrl:remove_observers()

-- init/start/stop each drive at least one state transition -> notification
assert(#states >= 3)

File: modules/py/test/core/py/test_observer.py
import sihd

# Observe a service's state transitions through the ServiceController bridge
core = sihd.core.Core("core")

ctrl = core.service_ctrl()
assert ctrl is not None

states = []
ctrl.add_observer(lambda c: states.append(c.state()))

assert core.init()
assert core.start()
assert core.stop()

ctrl.remove_observers()

# init/start/stop each drive at least one state transition -> notification
assert len(states) >= 3

print("observer tests passed")
